mithril-common/src/cardano_block_scanner/chain_reader_block_streamer.rs

```diff
@@ -0,0 +1,320 @@
+                        return Ok(Some(chain_scanned_blocks));
+                    }
+                }
+                Some(BlockStreamerNextAction::SkipToNextAction) => {
+                    continue;
+                }
+                None => {
+                    if roll_forwards.is_empty() {
+                        return Ok(None);
+                    } else {
+                        chain_scanned_blocks = ChainScannedBlocks::RollForwards(roll_forwards);
+                        return Ok(Some(chain_scanned_blocks));
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl ChainReaderBlockStreamer {
+    /// Factory
+    pub async fn try_new(
+        chain_reader: Arc<Mutex<dyn ChainBlockReader>>,
+        from: Option<ChainPoint>,
+        until: BlockNumber,
+        logger: Logger,
+    ) -> StdResult<Self> {
+        let from = from.unwrap_or(ChainPoint::origin());
+        {
+            let mut chain_reader_inner = chain_reader.try_lock()?;
+            chain_reader_inner.set_chain_point(&from).await?;
+        }
+        Ok(Self {
+            chain_reader,
+            from,
+            until,
+            max_roll_forwards_per_poll: MAX_ROLL_FORWARDS_PER_POLL,
+            logger,
+        })
+    }
+
+    async fn get_next_chain_block_action(&self) -> StdResult<Option<BlockStreamerNextAction>> {
+        let mut chain_reader = self.chain_reader.try_lock()?;
+        match chain_reader.get_next_chain_block().await? {
+            Some(ChainBlockNextAction::RollForward { parsed_block }) => {
+                if parsed_block.block_number >= self.until {
+                    debug!(
+                        self.logger,
+                        "ChainReaderBlockStreamer received a RollForward above threshold block number ({})",
+                        parsed_block.block_number
+                    );
+                    Ok(None)
+                } else {
+                    debug!(
+                        self.logger,
+                        "ChainReaderBlockStreamer received a RollForward below threshold block number ({})",
+                        parsed_block.block_number
+                    );
+                    Ok(Some(BlockStreamerNextAction::ChainBlockNextAction(
+                        ChainBlockNextAction::RollForward { parsed_block },
+                    )))
+                }
+            }
+            Some(ChainBlockNextAction::RollBackward { rollback_point }) => {
+                debug!(
+                    self.logger,
+                    "ChainReaderBlockStreamer received a RollBackward({rollback_point:?})"
+                );
+                let block_streamer_next_action = if rollback_point == self.from {
+                    BlockStreamerNextAction::SkipToNextAction
+                } else {
+                    BlockStreamerNextAction::ChainBlockNextAction(
+                        ChainBlockNextAction::RollBackward { rollback_point },
```

Comment:
Do we need to create a new Action ? 
Can we reuse the one that match the condition (I'm not sure this is any clearer) ?

Reply:
This new type expresses the fact that there is a case where we need to skip an action. 
This is something specific to the implementation of the the block streamer and does not exist in the `ChainReader`.
As we already use an `Option` for the case where the stream has ended, there is no other way to do it.